File: MathCore.DSP/Filters/ChebyshevBandStop.cs
﻿using System.ComponentModel;

using static System.Math;
using static MathCore.Polynom.Array;

namespace MathCore.DSP.Filters;

public class ChebyshevBandStop : ChebyshevFilter
{
    /// <summary>Формирование спецификации фильтра</summary>
    /// <param name="dt">Период дискретизации сигнала</param>
    /// <param name="fpl">Частота нижней границы полосы пропускания</param>
    /// <param name="fsl">Частота нижней границы полосы заграждения</param>
    /// <param name="fsh">Частота верхней границы полосы заграждения</param>
    /// <param name="fph">Частота верхней границы полосы пропускания</param>
    /// <param name="Gp">Уровень АЧХ в полосе пропускания (в единицах 0..1)</param>
    /// <param name="Gs">Неравномерность АЧХ в полосе заграждения (в единицах 0..1)</param>
    /// <returns>Спецификация фильтра</returns>
    private static Specification GetSpecification(
        double dt,
        double fpl,
        double fsl,
        double fsh,
        double fph,
        double Gp,
        double Gs)
    {
        if (Gp <= Gs)
            throw new ArgumentOutOfRangeException(
                nameof(Gp), Gp, $"Уровень АЧХ в полосе пропускания Gp={Gp} был меньше, либо равен уровню АЧХ в полосе заграждения Gs={Gs}");

        var Fpl = ToDigitalFrequency(fpl, dt);
        var Fsl = ToDigitalFrequency(fsl, dt);
        var Fsh = ToDigitalFrequency(fsh, dt);
        var Fph = ToDigitalFrequency(fph, dt);

        var Wpl = Consts.pi2 * Fpl;
        var Wsl = Consts.pi2 * Fsl;
        var Wsh = Consts.pi2 * Fsh;
        var Wph = Consts.pi2 * Fph;

        var Wc = Wsl * Wsh;
        var dW = Wsh - Wsl;

        // Выбор опорной частоты
        // Если   Wp / Wph > Wpl
        // то есть      Wc > Wpl*Wph
        // то есть Wsl*Wsh > Wpl*Wph
        // то есть центральная частота по границам подавления > центральной частоты по границам пропускания
        // то выбираем в качестве опорной частоты выбираем верхнюю границу пропускания
        // иначе, выбираем нижнюю границу пропускания
        var Wp = Wc / Wph > Wpl
            ? Wph
            : Wpl;
        var Fp = Abs(dW * Wp / (Wc - Wp.Pow2())) / Consts.pi2;
        const double Fs = 1 / Consts.pi2;

        // Для передачи информации о граничных частотах в спецификацию аналогвого прототипа перечситываем частоты цифрового фильтра обратно
        var fp = ToAnalogFrequency(Fp, dt);
        var fs = ToAnalogFrequency(Fs, dt);

        return new(dt, fp, fs, Gp, Gs);
    }

    /// <summary>Расчёт коэффициентов полиномов числителя из знаменателя передаточной функции фильтра</summary>
    /// <param name="fpl">Нижняя частота полосы пропускания</param>
    /// <param name="fsl">Нижняя частота полосы подавления</param>
    /// <param name="fsh">Верхняя частота полосы подавления</param>
    /// <param name="fph">Верхняя частота полосы пропускания</param>
    /// <param name="Spec">Спецификация фильтра</param>
    /// <returns>Кортеж, содержащий массивы A - коэффициенты полинома знаменателя и B - коэффициенты полинома числителя</returns>
    private static (double[] A, double[] B) InitializeI(double fpl, double fsl, double fsh, double fph, Specification Spec)
    {
        // Пересчитываем аналоговые частоты полосы заграждения в цифровые
        var dt = Spec.dt;

        var Wpl = Consts.pi2 * ToDigitalFrequency(fpl, dt);
        var Wsl = Consts.pi2 * ToDigitalFrequency(fsl, dt);
        var Wsh = Consts.pi2 * ToDigitalFrequency(fsh, dt);
        var Wph = Consts.pi2 * ToDigitalFrequency(fph, dt);

        var Wc = Wsl * Wsh;
        var dW = Wsh - Wsl;
        var sqrtWc = Wc.Sqrt();
        var Wp = Wc / Wph > Wpl
            ? Wph
            : Wpl;
        var W0 = Abs(dW * Wp / (Wc - Wp.Pow2()));

        var N = (int)Ceiling(arcch(Spec.kEps) / arcch(Spec.kW));
        var poles = GetNormedPolesI(N, Spec.EpsP, W0);

        // Переносим нули и полюса аналогового нормированного ФНЧ в полосы ПЗФ
        var pzf_zeros = Enumerable.Range(0, 2 * N).Select(i => i % 2 == 0 ? Complex.ImValue(sqrtWc) : Complex.ImValue(-sqrtWc));
        var pzf_poles = TransformToBandStopW(poles, Wsl, Wsh);

        // Преобразуем аналоговые нули и полюса в нули и полюса цифрового фильтра с помощью Z-преобразования
        var z_zeros = ToZArray(pzf_zeros, dt);
        var z_poles = ToZArray(pzf_poles, dt);

        // Вычисляем коэффициент нормировки фильтра на нулевой частоте 
        var g_norm = (N.IsOdd() ? 1 : Spec.Gp)
            / (z_zeros.Multiply(z => 1 - z) / z_poles.Multiply(z => 1 - z)).Abs;
        // Определяем массивы нулей коэффициентов полиномов знаменателя и числителя
        var B = GetCoefficientsInverted(z_zeros).ToArray(b => b.Re * g_norm);
        var A = GetCoefficientsInverted(z_poles).ToRe();

        return (A, B);
    }

    private static (double[] A, double[] B) InitializeII(double fpl, double fsl, double fsh, double fph, Specification Spec)
    {
        // Пересчитываем аналоговые частоты полосы заграждения в цифровые
        var dt = Spec.dt;

        var Wpl = Consts.pi2 * ToDigitalFrequency(fpl, dt);
        var Wsl = Consts.pi2 * ToDigitalFrequency(fsl, dt);
        var Wsh = Consts.pi2 * ToDigitalFrequency(fsh, dt);
        var Wph = Consts.pi2 * ToDigitalFrequency(fph, dt);

        var Wc = Wsl * Wsh;
        var dW = Wsh - Wsl;
        var Wp = Wc / Wph > Wpl
            ? Wph
            : Wpl;
        var W0 = Abs(dW * Wp / (Wc - Wp.Pow2()));

        var N = (int)Ceiling(arcch(Spec.kEps) / arcch(Spec.kW));
        var (zeros, poles) = GetNormedPolesII(N, Spec.EpsS, W0);

        // Переносим нули и полюса аналогового нормированного ФНЧ в полосы ПЗФ
        var sqrtWc = Wc.Sqrt();
        var is_even = N.IsEven();
        var pzf_zeros = is_even 
            ? TransformToBandStopW(zeros, Wsl, Wsh)
            : TransformToBandStopW(zeros, Wsl, Wsh).AppendLast((0, +sqrtWc), (0, -sqrtWc));
        var pzf_poles = TransformToBandStopW(poles, Wsl, Wsh);

        // Преобразуем аналоговые нули и полюса в нули и полюса цифрового фильтра с помощью Z-преобразования
        var z_zeros = ToZArray(pzf_zeros, dt);
        var z_poles = ToZArray(pzf_poles, dt);

        // Вычисляем коэффициент нормировки фильтра на нулевой частоте 
        var g_norm = 1 / (z_zeros.Multiply(z => 1 - z) / z_poles.Multiply(z => 1 - z)).Abs;

        // Определяем массивы нулей коэффициентов полиномов знаменателя и числителя
        var B = GetCoefficientsInverted(z_zeros).ToArray(b => b.Re * g_norm);
        var A = GetCoefficientsInverted(z_poles).ToRe();

        return (A, B);
    }

    private static (double[] A, double[] B) InitializeIICorrected(double fpl, double fsl, double fsh, double fph, Specification Spec)
    {
        // Пересчитываем аналоговые частоты полосы заграждения в цифровые
        var dt = Spec.dt;

        var Wpl = Consts.pi2 * ToDigitalFrequency(fpl, dt);
        var Wsl = Consts.pi2 * ToDigitalFrequency(fsl, dt);
        var Wsh = Consts.pi2 * ToDigitalFrequency(fsh, dt);
        var Wph = Consts.pi2 * ToDigitalFrequency(fph, dt);

        var Wc = Wsl * Wsh;
        var dW = Wsh - Wsl;
        var sqrt_wc = Wc.Sqrt();
        var Wp = Wc / Wph > Wpl
            ? Wph
            : Wpl;
        var W0 = Abs(dW * Wp / (Wc - Wp.Pow2()));

        var N = (int)Ceiling(arcch(Spec.kEps) / arcch(1 / W0));
        var (zeros, poles) = GetNormedPolesII(N, Spec.EpsS);

        // Переносим нули и полюса аналогового нормированного ФНЧ в полосы ПЗФ
        var is_even = N.IsEven();
        var pzf_zeros = is_even
            ? TransformToBandStopW(zeros, Wsl, Wsh)
            : TransformToBandStopW(zeros, Wsl, Wsh).AppendLast((0, +sqrt_wc), (0, -sqrt_wc));
        var pzf_poles = TransformToBandStopW(poles, Wsl, Wsh);

        // Преобразуем аналоговые нули и полюса в нули и полюса цифрового фильтра с помощью Z-преобразования
        var z_zeros = ToZArray(pzf_zeros, dt);
        var z_poles = ToZArray(pzf_poles, dt);

        // Вычисляем коэффициент нормировки фильтра на нулевой частоте 
        var re_to_im = z_zeros.Multiply(z => 1 - z) / z_poles.Multiply(z => 1 - z);
        var g_norm = 1 / re_to_im.Re;

        // Определяем массивы нулей коэффициентов полиномов знаменателя и числителя
        var B = GetCoefficientsInverted(z_zeros).ToArray(b => b.Re * g_norm);
        var A = GetCoefficientsInverted(z_poles).ToRe();

        return (A, B);
    }

    /// <summary>Инициализация нового эллиптического полосозаграждающего фильтра (ПЗФ)</summary>
    /// <param name="dt">Период дискретизации цифрового сигнала</param>
    /// <param name="fpl">Нижняя граница полосы пропускания</param>
    /// <param name="fsl">Нижняя граница полосы подавления</param>
    /// <param name="fsh">Верхняя граница полосы подавления</param>
    /// <param name="fph">Верхняя граница полосы пропускания</param>
    /// <param name="Gp">Уровень АЧХ в полосе пропускания (по умолчанию -1дБ)</param>
    /// <param name="Gs">Уровень АЧХ в полосе подавления (по умолчанию -40дБ)</param>
    /// <param name="Type">Тип фильтра I или II</param>
    public ChebyshevBandStop(
        double dt,
        double fpl,
        double fsl,
        double fsh,
        double fph,
        double Gp = 0.89125093813374556,
        double Gs = 0.01, 
        ChebyshevType Type = ChebyshevType.I)
        : this(fpl, fsl, fsh, fph, GetSpecification(dt, fpl, fsl, fsh, fph, Gp, Gs), Type) { }

    /// <summary>Инициализация нового эллиптического полосозаграждающего фильтра (ПЗФ)</summary>
    /// <param name="fpl">Нижняя граница полосы пропускания</param>
    /// <param name="fsl">Нижняя граница полосы подавления</param>
    /// <param name="fsh">Верхняя граница полосы подавления</param>
    /// <param name="fph">Верхняя граница полосы пропускания</param>
    /// <param name="Spec">Спецификация фильтра</param>
    /// <param name="Type">Тип фильтра I или II</param>
    private ChebyshevBandStop(double fpl, double fsl, double fsh, double fph, Specification Spec, ChebyshevType Type)
        : this(Type switch
        {
            ChebyshevType.I => InitializeI(fpl, fsl, fsh, fph, Spec),
            ChebyshevType.II => InitializeII(fpl, fsl, fsh, fph, Spec),
            ChebyshevType.IICorrected => InitializeIICorrected(fpl, fsl, fsh, fph, Spec),
            _ => throw new InvalidEnumArgumentException(nameof(Type), (int)Type, typeof(ChebyshevType))
        }, Spec, Type) { }

    /// <summary>Инициализация нового эллиптического полосозаграждающего фильтра (ПЗФ)</summary>
    /// <param name="Polynoms">Кортеж с коэффициентами полиномов знаменателя и числителя функции фильтра</param>
    /// <param name="Spec">Спецификация фильтра</param>
    /// <param name="Type">Тип фильтра I или II</param>
    private ChebyshevBandStop((double[] A, double[] B) Polynoms, Specification Spec, ChebyshevType Type) 
        : this(Polynoms.B, Polynoms.A, Spec, Type) { }

    /// <summary>Инициализация нового эллиптического полосозаграждающего фильтра (ПЗФ)</summary>
    /// <param name="B">Коэффициенты полинома числителя</param>
    /// <param name="A">Коэффициенты полинома знаменателя</param>
    /// <param name="Spec">Спецификация фильтра</param>
    /// <param name="Type">Тип фильтра I или II</param>
    private ChebyshevBandStop(double[] B, double[] A, Specification Spec, ChebyshevType Type) : base(B, A, Spec, Type) { }
}

File: MathCore.DSP/Filters/Python/filters/chebyshev_bandstop.py
# Полосозаграждающий фильтр Чебышева (порт C# ChebyshevBandStop)
import numpy as np
from math import ceil, pi, fabs

from .chebyshev_filter import ChebyshevFilter, ChebyshevType
from .analog_based_filter import AnalogBasedFilter, Specification
from .digital_filter import DigitalFilter

pi2 = 2.0 * pi


class ChebyshevBandStop(ChebyshevFilter):
    """Цифровой полосозаграждающий фильтр Чебышева (ПЗФ)."""

    @staticmethod
    def _get_specification(dt: float, fpl: float, fsl: float, fsh: float, fph: float,
                           Gp: float, Gs: float) -> Specification:
        """Спецификация ПЗФ через эквивалентный ФНЧ-прототип."""
        Fpl = DigitalFilter.to_digital_frequency(fpl, dt)
        Fsl = DigitalFilter.to_digital_frequency(fsl, dt)
        Fsh = DigitalFilter.to_digital_frequency(fsh, dt)
        Fph = DigitalFilter.to_digital_frequency(fph, dt)

        Wpl, Wsl = pi2 * Fpl, pi2 * Fsl
        Wsh, Wph = pi2 * Fsh, pi2 * Fph

        Wc = Wsl * Wsh
        dW = Wsh - Wsl

        Wp = Wph if (Wc / Wph > Wpl) else Wpl
        Fp = fabs(dW * Wp / (Wc - Wp ** 2)) / pi2
        Fs = 1.0 / pi2

        fp = DigitalFilter.to_analog_frequency(Fp, dt)
        fs = DigitalFilter.to_analog_frequency(Fs, dt)
        return Specification(dt, fp, fs, Gp, Gs)

    @staticmethod
    def _common_params(fpl: float, fsl: float, fsh: float, fph: float, dt: float):
        """Вычисляет общие параметры ПЗФ (Wc, dW, sqrt_wc, W0, Wp)."""
        Wpl = pi2 * DigitalFilter.to_digital_frequency(fpl, dt)
        Wsl = pi2 * DigitalFilter.to_digital_frequency(fsl, dt)
        Wsh = pi2 * DigitalFilter.to_digital_frequency(fsh, dt)
        Wph = pi2 * DigitalFilter.to_digital_frequency(fph, dt)

        Wc = Wsl * Wsh
        dW = Wsh - Wsl
        sqrt_wc = np.sqrt(Wc)

        Wp_ref = Wph if (Wc / Wph > Wpl) else Wpl
        W0 = fabs(dW * Wp_ref / (Wc - Wp_ref ** 2))
        return Wsl, Wsh, Wc, dW, sqrt_wc, W0

    @staticmethod
    def _normalize_bs(z_zeros, z_poles, N, spec):
        """Нормировка ПЗФ по постоянной составляющей."""
        gain_zeros = np.prod([1.0 - z for z in z_zeros])
        gain_poles = np.prod([1.0 - z for z in z_poles])
        k0 = 1.0 if N % 2 != 0 else spec.Gp
        return k0 / abs(gain_zeros / gain_poles)

    @staticmethod
    def _initialize_i(fpl: float, fsl: float, fsh: float, fph: float, spec: Specification):
        """Коэффициенты ПЗФ Чебышева I рода."""
        dt = spec.dt
        Wsl, Wsh, Wc, dW, sqrt_wc, W0 = ChebyshevBandStop._common_params(
            fpl, fsl, fsh, fph, dt)

        N = int(ceil(ChebyshevFilter.arcch(spec.kEps) / ChebyshevFilter.arcch(spec.kW)))

        poles = ChebyshevFilter.get_normed_poles_i(N, spec.EpsP, W0)

        pzf_zeros = [complex(0.0, sqrt_wc if i % 2 == 0 else -sqrt_wc)
                     for i in range(2 * N)]
        pzf_poles = AnalogBasedFilter.transform_to_band_stop_w(poles, Wsl, Wsh)

        z_zeros = DigitalFilter.to_z_array(pzf_zeros, dt)
        z_poles = DigitalFilter.to_z_array(pzf_poles, dt)

        g_norm = ChebyshevBandStop._normalize_bs(z_zeros, z_poles, N, spec)
        return np.poly(z_poles).real, np.poly(z_zeros).real * g_norm

    @staticmethod
    def _initialize_ii(fpl: float, fsl: float, fsh: float, fph: float, spec: Specification):
        """Коэффициенты ПЗФ Чебышева II рода."""
        dt = spec.dt
        Wsl, Wsh, Wc, dW, sqrt_wc, W0 = ChebyshevBandStop._common_params(
            fpl, fsl, fsh, fph, dt)

        N = int(ceil(ChebyshevFilter.arcch(spec.kEps) / ChebyshevFilter.arcch(spec.kW)))

        zeros, poles = ChebyshevFilter.get_normed_poles_ii(N, spec.EpsS, W0)

        pzf_poles = AnalogBasedFilter.transform_to_band_stop_w(poles, Wsl, Wsh)
        pzf_zeros = list(AnalogBasedFilter.transform_to_band_stop_w(zeros, Wsl, Wsh))
        if N % 2 != 0:
            pzf_zeros += [complex(0.0, sqrt_wc), complex(0.0, -sqrt_wc)]

        z_zeros = DigitalFilter.to_z_array(pzf_zeros, dt)
        z_poles = DigitalFilter.to_z_array(pzf_poles, dt)

        gain_zeros = np.prod([1.0 - z for z in z_zeros])
        gain_poles = np.prod([1.0 - z for z in z_poles])
        g_norm = 1.0 / abs(gain_zeros / gain_poles)

        return np.poly(z_poles).real, np.poly(z_zeros).real * g_norm

    @staticmethod
    def _initialize_ii_corrected(fpl: float, fsl: float, fsh: float, fph: float,
                                 spec: Specification):
        """Коэффициенты ПЗФ Чебышева II рода с коррекцией частот."""
        dt = spec.dt
        Wsl, Wsh, Wc, dW, sqrt_wc, W0 = ChebyshevBandStop._common_params(
            fpl, fsl, fsh, fph, dt)

        N = int(ceil(ChebyshevFilter.arcch(spec.kEps) / ChebyshevFilter.arcch(1.0 / W0)))

        zeros, poles = ChebyshevFilter.get_normed_poles_ii(N, spec.EpsS)

        pzf_poles = AnalogBasedFilter.transform_to_band_stop_w(poles, Wsl, Wsh)
        pzf_zeros = list(AnalogBasedFilter.transform_to_band_stop_w(zeros, Wsl, Wsh))
        if N % 2 != 0:
            pzf_zeros += [complex(0.0, sqrt_wc), complex(0.0, -sqrt_wc)]

        z_zeros = DigitalFilter.to_z_array(pzf_zeros, dt)
        z_poles = DigitalFilter.to_z_array(pzf_poles, dt)

        gain_zeros = np.prod([1.0 - z for z in z_zeros])
        gain_poles = np.prod([1.0 - z for z in z_poles])
        g_norm = 1.0 / (gain_zeros / gain_poles).real

        return np.poly(z_poles).real, np.poly(z_zeros).real * g_norm

    def __init__(self, dt: float, fpl: float, fsl: float, fsh: float, fph: float,
                 Gp: float = 0.891250938, Gs: float = 0.01,
                 filter_type: ChebyshevType = ChebyshevType.I):
        """
        Создаёт ПЗФ Чебышева.

        Параметры
        ---------
        dt          : период дискретизации (с)
        fpl         : нижняя граница полосы пропускания (Гц)
        fsl         : нижняя граница полосы заграждения (Гц)
        fsh         : верхняя граница полосы заграждения (Гц)
        fph         : верхняя граница полосы пропускания (Гц)
        Gp          : коэффициент передачи на границе пропускания
        Gs          : коэффициент передачи на границе заграждения
        filter_type : тип фильтра
        """
        spec = ChebyshevBandStop._get_specification(dt, fpl, fsl, fsh, fph, Gp, Gs)
        init = {
            ChebyshevType.I:
                lambda s: ChebyshevBandStop._initialize_i(fpl, fsl, fsh, fph, s),
            ChebyshevType.II:
                lambda s: ChebyshevBandStop._initialize_ii(fpl, fsl, fsh, fph, s),
            ChebyshevType.IICorrected:
                lambda s: ChebyshevBandStop._initialize_ii_corrected(fpl, fsl, fsh, fph, s),
        }
        A, B = init[filter_type](spec)
        super().__init__(B, A, spec, filter_type)
